sw/device/lib/dif/dif_spi_device.h

```diff
@@ -0,0 +1,233 @@
+  dif_spi_device_bit_order_t tx_order;
+  dif_spi_device_bit_order_t rx_order;
+  uint16_t rx_fifo_timeout;  // uint12_t
+  uint16_t rx_fifo_len;
+  uint16_t tx_fifo_len;
+} dif_spi_device_config_t;
+
+/**
+ * State for a particular SPI device.
+ *
+ * Its member variables should be considered private, and are only provided so
+ * that callers can allocate it.
+ */
+typedef struct dif_spi_device {
+  mmio_region_t base_addr;
+  uint16_t rx_fifo_base;
+  uint16_t rx_fifo_len;
+  uint16_t tx_fifo_base;
+  uint16_t tx_fifo_len;
+} dif_spi_device_t;
+
+/**
+ * Initializes a SPI device with the given configuration.
+ *
+ * @param config the configuration to initialize with.
+ * @param spi out param for the initialized device.
+ * @return the result of the operation.
+ */
+dif_spi_device_result_t dif_spi_device_init(dif_spi_device_config_t config,
+                                            dif_spi_device_t *spi);
+
+/**
+ * Issues an "abort" to the given SPI device, causing all in-progress IO to
+ *
+ * @param spi a SPI device.
+ * @return the result of the operation.
+ */
+dif_spi_device_result_t dif_spi_device_abort(const dif_spi_device_t *spi);
+
+/**
+ * Issues an "abort" to the given SPI device, causing all in-progress IO to
+ *
+ * @param spi a SPI device.
+ * @return the result of the operation.
+ */
+dif_spi_device_result_t dif_spi_device_abort(const dif_spi_device_t *spi);
+
+/**
+ * Checks whether a particular interrupt bit has been set.
+ *
+ * @param spi a SPI device.
+ * @param type which IRQ type to check.
+ * @param flag out param for the interrupt bit.
+ * @return the result of the operation.
+ */
+dif_spi_device_result_t dif_spi_device_irq_get(const dif_spi_device_t *spi,
+                                               dif_spi_device_irq_type_t type,
+                                               bool *flag_out);
+
+/**
+ * Clears all active interrupt bits.
+ *
+ * @param spi a SPI device.
+ * @return the result of the operation.
+ */
+dif_spi_device_result_t dif_spi_device_irq_clear_all(
+    const dif_spi_device_t *spi);
+
+/**
+ * Toggles a particular interrupt enable bit to enabled or disabled.
+ *
+ * @param spi a SPI device.
+ * @param type which IRQ type to toggle.
+ * @param state the state to update the bit to.
+ * @return the result of the operation.
+ */
+dif_spi_device_result_t dif_spi_device_irq_toggle(
+    const dif_spi_device_t *spi, dif_spi_device_irq_type_t type,
+    dif_spi_device_irq_state_t state);
+
+/**
+ * Forces a particular IRQ type to fire.
+ *
+ * @param spi a SPI device.
+ * @param type which IRQ type to fire.
+ * @return the result of the operation.
+ */
+dif_spi_device_result_t dif_spi_device_irq_force(
+    const dif_spi_device_t *spi, dif_spi_device_irq_type_t type);
+
+/**
+ * Sets up the "FIFO level" (that is, number of bytes present in a particular
+ * FIFO) at which the TxLevel and RxLevel IRQs will fire.
+ *
+ * For the reciept side, when the FIFO overflows |rx_level|, an interrupt is
```

Comment:
Here and all other places, instead of `|`, use `@p param` format.

Reply:
It's not clear that we're doing `@p param`, from the looks of comments on Sam's issue.

Reply:
I thought that in function descriptions `@p` was not controversial, but I might be wrong.

Reply:
Sam looked into it in #1849, it seems like what `@p` parses is kind of inconsistent, so we're sticking with `` `...` `` for now.

Reply:
With @lenary #1881 change these 3 witch statements could then be combined, I think. (when it lands)